### PR TITLE
vscode/haskell: init with Haskell IDE Engine support

### PR DESCRIPTION
**Pull request by roberth**
Adds:
 - Haskell IDE Engine integration
 - Syntax highlighting

### CODE REVIEW

**Comment by rycee** [comment]
Thanks for the contribution! Unfortunately this module is a bit problematic in that it includes several packages that will have to be kept up to date. I'm very reluctant to include packages in Home Manager.

File: modules/misc/news.nix [hunk 1]
```diff
@@ -998,6 +998,17 @@ in
           A new service is available: 'services.mpdris2'.
         '';
       }
+
+      {
+        time = "2019-03-17T13:37:00+00:00";
+        condition = config.programs.vscode.enable;
+        message = ''
+          A new module is available: 'services.mpdris2'.
```
Comment (rycee) [inline modules/misc/news.nix]: I think you want `programs.vscode.haskell` here.

File: modules/programs/vscode/haskell.nix [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  };
+  hie-nix = import cfg.hieNixSource hieNixArgs;
+
+in
+{
+  options.programs.vscode.haskell.enable = lib.mkEnableOption "Haskell integration for Visual Studio Code";
+
+  options.programs.vscode.haskell.hieNixSource = lib.mkOption {
+    type = lib.types.path;
+    default = defaultHieNixSource;
```
Comment (rycee) [inline modules/programs/vscode/haskell.nix]: I think there should be no default here to force the user to choose a hie package themselves.

File: modules/programs/vscode/haskell.nix [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  hieNixArgs = lib.optionalAttrs (!cfg.hieNixUsePinnedPkgs) {
+    inherit pkgs;
+  };
+  hie-nix = import cfg.hieNixSource hieNixArgs;
+
+in
+{
+  options.programs.vscode.haskell.enable = lib.mkEnableOption "Haskell integration for Visual Studio Code";
+
+  options.programs.vscode.haskell.hieNixSource = lib.mkOption {
```
Comment (rycee) [inline modules/programs/vscode/haskell.nix]: Can break out the common attribute name parts. E.g. to have
```
options.programs.vscode.haskell = {
  enable = …;
  hie = {
    nixSource = …;
    executablePath = …;
    nixUsePinnedPkgs = …;
  };
};
```

**Comment by roberth** [comment]
I understand your concern about packages. I'll remove the default hie-nix source. Is it ok to include the vscode extensions if I put a comment there that says that these packages are placeholders that will be removed when they've been in Nixpkgs for a year? That way they will not have to be maintained here.

File: modules/programs/vscode/haskell.nix [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    '';
+    example = lib.literalExample ''
+      ~/src/haskell-ide-engine
+    '';
+    description = ''
+      A version of the <link xlink:href="https://github.com/domenkozar/hie-nix/">hie-nix sources</link>.
+    '';
+  };
+
+  options.programs.vscode.haskell.hieExecutablePath = lib.mkOption {
```
Comment (rycee) [inline modules/programs/vscode/haskell.nix]: Seems to me like this is the option that matters. Perhaps one could simply remove the `hieNixSource` and `hieNixUsePinnedPkgs` options? Setting up the package seems like something the user should do in an overlay or `packageOverrides`.

**Comment by rycee** [comment]
Are the vscode extensions on their way into Nixpkgs? I would suggest to wait with merging this module until they are available in Nixpkgs. The fallback packages won't be necessary then, people tracking Nixpkgs master should update regularly 😀

File: modules/programs/vscode/haskell.nix (hunk 3 above)
Comment (rycee) [reply modules/programs/vscode/haskell.nix]: Note, I removed the `hie` prefix and instead made it an extra attribute level. Generally I find the option names turn out nicer if they don't include "namespace" concerns.

Comment (roberth) [reply modules/programs/vscode/haskell.nix]: Only hie.executablePath remains, after the changes you suggest, or it'd have to be hie.hie-nixSource or hie.hieNixSource.

**Comment by roberth** [comment]
I can upstream the extensions in Nixpkgs first.

**Comment by roberth** [comment]
I don't like that the burden of figuring out how to get and build hie-nix will be on the user. They'll expect to find it in Nixpkgs, but packaging it there is simply not feasible and the hie-nix repo will be the only method of installation in the future.

It just seems to defeat the purpose of this module for almost no reason.

**Comment by eadwu** [comment]
Does the extension work for you? I'm only had partial success with it. The furthest I've gotten to setting up this extension was up to `cabal configure`. Even if I hack past that part, I get dependency problems. It also seems like the `formattingProvider` setting is broken (or maybe it isn't but the `none` value still uses `brittany`).

**Comment by roberth** [comment]
I'm using it with Stack and its Nix integration. I was using brittany anyway so didn't notice any problems :). I should blog about my project setup + any issues when this is merged.

**Comment by terlar** [comment]
I can second that, I have it working with both cabal and stack projects together with britanny as formatter. I am using it together with emacs and eglot though

**Comment by roberth** [comment]
@rycee The extensions are in Nixpkgs master now and I've made some changes to the module here. Anything else that needs to be done?

**Comment by rycee** [comment]
@roberth Looks good. I made a few mostly cosmetic changes and pushed to your branch. Please give it a try and if everything works as before I'd be happy to merge as-is.

**Comment by roberth** [comment]
It works. Ready for merge.

**Comment by worldofpeace** [comment]
@rycee :ping_pong:

**Comment by rycee** [comment]
Thanks for the reminder 🙂

Rebased to master in c5f230e68219252ecda608fcd1f5e85869687d09.

**Comment by teto** [comment]
hie-nix is now superseeded by hies-all so it would be nice to fix the error message requesting to install the hie-overlay.

**Pull request closed without merging**